ebmlite/util.py

```diff
@@ -11,20 +11,102 @@
     `toXml()` and `xml2ebml()`.
 """
 __author__ = "David Randall Stokes, Connor Flanigan"
-__copyright__ = "Copyright 2020, Mide Technology Corporation"
+__copyright__ = "Copyright 2021, Mide Technology Corporation"
 __credits__ = "David Randall Stokes, Connor Flanigan, Becker Awqatty, Derek Witt"
 
-__all__ = ['toXml', 'xml2ebml', 'loadXml', 'pprint']
-
-from io import StringIO
+__all__ = ['createID', 'validateID', 'toXml', 'xml2ebml', 'loadXml', 'pprint']
 
 import ast
 from base64 import b64encode, b64decode
+from io import StringIO
+import struct
 import sys
 import tempfile
 from xml.etree import ElementTree as ET
 
-from . import core, encoding
+from . import core, encoding, decoding
+
+# ==============================================================================
+#
+# ==============================================================================
+
+
+def createID(schema, idClass, exclude=(), minId=0x80, maxId=0x1FFFFFFF, count=1):
+    """ Generate unique EBML IDs. Primarily intended for use 'offline' by
+        humans creating EBML schemata.
+
+        @param schema: The `Schema` in which the new IDs must coexist.
+        @param idClass: The EBML class of ID, one of (case-insensitive):
+            * `'a'`: Class A (1 octet, base 0x8X)
+            * `'b'`: Class B (2 octets, base 0x4000)
+            * `'c'`: Class C (3 octets, base 0x200000)
+            * `'d'`: Class D (4 octets, base 0x10000000)
+        @param exclude: A list of additional IDs to avoid.
+        @param minId: The minimum ID value, within the ID class' range.
+        @param maxId: The maximum ID value, within the ID class' range.
+        @param count: The maximum number of IDs to generate. The result may be
+            fewer than specified if too few meet the given criteria.
+        @return: A list of EBML IDs that match the given criteria.
+    """
+    ranges = dict(A=(0x80, 0xFF),
+                  B=(0x4000, 0x7FFF),
+                  C=(0x200000, 0x3FFFFF),
+                  D=(0x10000000, 0x1FFFFFFF))
```

Comment:
was looking at the EBML spec the other day; seems that the range of valid Element ID's *excludes* any with all 0's or 1's in the data. this means that the actual ranges would be as copied below:
```python
ranges = dict(
    A=(0x81, 0xFE),
    B=(0x4001, 0x7FFE),
    C=(0x200001, 0x3FFFFE),
    D=(0x10000001, 0x1FFFFFFE),
)

```

reference: https://github.com/ietf-wg-cellar/ebml-specification/blob/master/specification.markdown#element-id

Reply:
Good catch! Addressing it now.